flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/DataFileWriter.java

```diff
@@ -97,172 +99,176 @@ public DataFilePathFactory pathFactory() {
+    }
 
-        protected void resetMeta() {
-            rowCount = 0;
-            minKey = null;
-            maxKey = null;
-            minSequenceNumber = Long.MAX_VALUE;
-            maxSequenceNumber = Long.MIN_VALUE;
-        }
+    private class KvFileWriter extends BaseFileWriter<KeyValue> {
+        private final int level;
+        private final RowDataSerializer keySerializer;
 
-        protected abstract KeyAndValueStats extractStats(Path path);
-    }
+        private FieldStatsCollector keyStatsCollector = null;
+        private FieldStatsCollector valueStatsCollector = null;
 
-    private class FileExtractingRollingFile extends DataRollingFile {
+        private BinaryRowData minKey = null;
+        private BinaryRowData maxKey = null;
+        private Long minSequenceNumber = null;
+        private Long maxSequenceNumber = null;
 
-        private FileExtractingRollingFile(int level) {
-            super(level);
+        public KvFileWriter(BulkWriter.Factory<KeyValue> writerFactory, Path path, int level)
+                throws IOException {
+            super(writerFactory, path);
+
+            this.level = level;
+            this.keySerializer = new RowDataSerializer(keyType);
+            if (fileStatsExtractor == null) {
+                this.keyStatsCollector = new FieldStatsCollector(keyType);
+                this.valueStatsCollector = new FieldStatsCollector(valueType);
+            }
         }
 
         @Override
-        protected KeyAndValueStats extractStats(Path path) {
-            FieldStats[] rawStats;
-            try {
-                rawStats = fileStatsExtractor.extract(path);
-            } catch (IOException e) {
-                throw new RuntimeException(e);
+        public void write(KeyValue kv) throws IOException {
+            super.write(kv);
+
+            if (fileStatsExtractor == null) {
+                keyStatsCollector.collect(kv.key());
+                valueStatsCollector.collect(kv.value());
             }
 
-            int numKeyFields = keyType.getFieldCount();
-            return new KeyAndValueStats(
-                    Arrays.copyOfRange(rawStats, 0, numKeyFields),
-                    Arrays.copyOfRange(rawStats, numKeyFields + 2, rawStats.length));
-        }
-    }
+            updateMinKey(kv);
+            updateMaxKey(kv);
 
-    private class StatsCollectingRollingFile extends DataRollingFile {
+            updateMinSeqNumber(kv);
+            updateMaxSeqNumber(kv);
+        }
 
-        private FieldStatsCollector keyStatsCollector;
-        private FieldStatsCollector valueStatsCollector;
+        private void updateMinKey(KeyValue kv) {
+            if (minKey == null) {
+                minKey = keySerializer.toBinaryRow(kv.key()).copy();
+            }
+        }
 
-        private StatsCollectingRollingFile(int level) {
-            super(level);
+        private void updateMaxKey(KeyValue kv) {
+            maxKey = keySerializer.toBinaryRow(kv.key()).copy();
```

Comment:
We should avoid copy per record.

Reply:
Yes, I just released this issue when I get the first round self-reviewing..